### PR TITLE
tmpfiles: refuse to chown()/chmod() files which are hardlinked, unles…

### PR DESCRIPTION
**Pull request by lnykryn**
…s protected_hardlinks sysctl is on

Let's add some extra safety.

Fixes: #7736

Cherry-picked from: 5579f85663d10269e7ac7464be6548c99cea4ada
Resolves: #1539758

### CODE REVIEW

**Comment by mrc0mmand** [comment]
Note: changing bug ID from BZ#1539758 (CLOSED WONTFIX) to BZ#1535135 (POST, RHEL 7.5+).

**Review by msekletar — APPROVED** [review]

**Comment by msekletar** [comment]
LGTM

File: src/tmpfiles/tmpfiles.c [hunk 1]
```diff
@@ -586,6 +586,39 @@ static int dir_cleanup(
...
+        static int cached = -1;
+        int r;
+
+        /* Check whether the fs.protected_hardlinks sysctl is on. If we can't determine it we assume its off, as that's
+         * what the upstream default is. */
+
+        if (cached >= 0)
+                return cached;
+
+        r = read_one_line_file("/proc/sys/fs/protected_hardlinks", &value);
```
Comment (mrc0mmand) [inline src/tmpfiles/tmpfiles.c:600]: Probably missing `fileio.h` include for this function.

```
src/tmpfiles/tmpfiles.c: In function 'dangerous_hardlinks':
src/tmpfiles/tmpfiles.c:600:9: error: implicit declaration of function 'read_one_line_file' [-Werror=implicit-function-declaration]
         r = read_one_line_file("/proc/sys/fs/protected_hardlinks", &value);
         ^
```

**Review by mrc0mmand — CHANGES_REQUESTED** [review]

**Comment by lnykryn** [comment]
Not needed in el7

**Pull request closed without merging**